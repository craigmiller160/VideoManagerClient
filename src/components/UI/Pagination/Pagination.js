import React from 'react';
import PropTypes from 'prop-types';
import classes from './Pagination.scss';
import { Pagination as ReactPagination, PaginationItem, PaginationLink } from 'reactstrap';

export const LEFT_ALIGN = 'left';
export const CENTER_ALIGN = 'center';
export const RIGHT_ALIGN = 'right';
const NUM_BUTTONS = 5;

export const getAlignClassName = (stringName) => {
    switch(stringName) {
        case LEFT_ALIGN:
            return classes.left;
        case CENTER_ALIGN:
            return classes.center;
        case RIGHT_ALIGN:
            return classes.right;
        default:
            throw new Error(`Invalid string align class name: ${stringName}`);
    }
};

const createPageButtons = (currentPage, totalPages, onClick) =>
    [...Array(totalPages).keys()]
        .filter((index) => index > currentPage - NUM_BUTTONS &&
            index < currentPage + NUM_BUTTONS)
        .map((index) => (
            <PaginationItem
                key={ index }
                active={ index === currentPage }
                onClick={ () => onClick(`${index}`) }
            >
                <PaginationLink style={ { zIndex: 0 } }>
                    { index + 1 }
                </PaginationLink>
            </PaginationItem>
        ));

const Pagination = (props) => {
    const { currentPage, totalPages, align, onClick } = props;
    const showPrevious = currentPage > 0;
    const showNext = currentPage < totalPages - 1;
    const classNames = [ classes.Pagination, getAlignClassName(align) ].join(' ');

    return (
        <ReactPagination className={ classNames }>
            {
                showPrevious &&
                    <>
                        <PaginationItem onClick={ () => onClick('<<') }>
                            <PaginationLink first />
                        </PaginationItem>
                        <PaginationItem onClick={ () => onClick('<') }>
                            <PaginationLink previous />
                        </PaginationItem>
                    </>
            }
            {
                createPageButtons(currentPage, totalPages, onClick)
            }
            {
                showNext &&
                    <>
                        <PaginationItem onClick={ () => onClick('>') }>
                            <PaginationLink next />
                        </PaginationItem>
                        <PaginationItem onClick={ () => onClick('>>') }>
                            <PaginationLink last />
                        </PaginationItem>
                    </>
            }
        </ReactPagination>
    );
};

Pagination.defaultProps = {
    align: LEFT_ALIGN,
    onClick: () => {}
};

Pagination.propTypes = {
    currentPage: PropTypes.number.isRequired,
    totalPages: PropTypes.number.isRequired,
    align: PropTypes.oneOf([LEFT_ALIGN, CENTER_ALIGN, RIGHT_ALIGN]),
    onClick: PropTypes.func
};

export default Pagination;