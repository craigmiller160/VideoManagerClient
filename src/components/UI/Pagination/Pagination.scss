.Pagination {
  display: flex;
  margin-top: 10px;
  &.left {
    justify-content: flex-start;
  }
  &.center {
    justify-content: center;
  }
  &.right {
    justify-content: flex-end;
  }
}