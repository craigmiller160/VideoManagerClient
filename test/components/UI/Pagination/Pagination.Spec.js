import React from 'react';
import Pagination, { RIGHT_ALIGN } from '../../../../src/components/UI/Pagination/Pagination';
import { mount } from 'enzyme';

const getProps = (currentPage) => ({
    currentPage,
    totalPages: 5,
    align: RIGHT_ALIGN
});

describe('Pagination', () => {
    it('should render correctly', () => {
        let clickedIndex = -1;
        const clickHandler = (index) => {
            clickedIndex = index;
        };

        const component = mount(
            <Pagination
                { ...getProps(1) }
                onClick={ clickHandler }
            />
        );
        expect(component).toMatchSnapshot();

        const paginationRoot = component.find('Pagination[tag="nav"]');
        expect(paginationRoot.hasClass('Pagination')).toEqual(true);
        expect(paginationRoot.hasClass('right')).toEqual(true);

        const paginationItems = component.find('PaginationItem');
        expect(paginationItems.length).toEqual(7);

        expect(paginationItems.at(0).find('PaginationLink').props().previous).toEqual(true);
        expect(paginationItems.at(6).find('PaginationLink').props().next).toEqual(true);
        expect(paginationItems.at(2).props().active).toEqual(true);

        paginationItems.at(3).simulate('click', {});
        expect(clickedIndex).toEqual('2');
        paginationItems.at(0).simulate('click', {});
        expect(clickedIndex).toEqual('<');
        paginationItems.at(6).simulate('click', {});
        expect(clickedIndex).toEqual('>');
    });

    it('should only render with next button', () => {
        const component = mount(
            <Pagination { ...getProps(0) } />
        );
        expect(component).toMatchSnapshot();

        const paginationItems = component.find('PaginationItem');
        expect(paginationItems.length).toEqual(6);

        expect(paginationItems.at(0).find('PaginationLink').props().previous).toBeUndefined();
        expect(paginationItems.at(5).find('PaginationLink').props().next).toEqual(true);
        expect(paginationItems.at(0).props().active).toEqual(true);
    });

    it('should only render with previous button', () => {
        const component = mount(
            <Pagination { ...getProps(4) } />
        );
        expect(component).toMatchSnapshot();

        const paginationItems = component.find('PaginationItem');
        expect(paginationItems.length).toEqual(6);

        expect(paginationItems.at(0).find('PaginationLink').props().previous).toEqual(true);
        expect(paginationItems.at(5).find('PaginationLink').props().next).toBeUndefined();
        expect(paginationItems.at(5).props().active).toEqual(true);
    });
});