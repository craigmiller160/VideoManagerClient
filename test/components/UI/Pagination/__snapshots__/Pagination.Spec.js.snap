// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should only render with next button 1`] = `
<nav
  aria-label="pagination"
  className="Pagination right"
>
  <ul
    className="pagination"
  >
    <li
      className="page-item active"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        1
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        2
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        3
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        4
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        5
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        aria-label="Next"
        className="page-link"
      >
        <span
          aria-hidden="true"
          key="caret"
        >
          »
        </span>
        <span
          className="sr-only"
          key="sr"
        >
          Next
        </span>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should only render with previous button 1`] = `
<nav
  aria-label="pagination"
  className="Pagination right"
>
  <ul
    className="pagination"
  >
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        aria-label="Previous"
        className="page-link"
      >
        <span
          aria-hidden="true"
          key="caret"
        >
          «
        </span>
        <span
          className="sr-only"
          key="sr"
        >
          Previous
        </span>
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        1
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        2
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        3
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        4
      </button>
    </li>
    <li
      className="page-item active"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        5
      </button>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should render correctly 1`] = `
<nav
  aria-label="pagination"
  className="Pagination right"
>
  <ul
    className="pagination"
  >
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        aria-label="Previous"
        className="page-link"
      >
        <span
          aria-hidden="true"
          key="caret"
        >
          «
        </span>
        <span
          className="sr-only"
          key="sr"
        >
          Previous
        </span>
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        1
      </button>
    </li>
    <li
      className="page-item active"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        2
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        3
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        4
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        5
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        aria-label="Next"
        className="page-link"
      >
        <span
          aria-hidden="true"
          key="caret"
        >
          »
        </span>
        <span
          className="sr-only"
          key="sr"
        >
          Next
        </span>
      </button>
    </li>
  </ul>
</nav>
`;
